contracts/utils/NoncesSemiAbstracted.sol

```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.20;
+
+import {Nonces} from "./Nonces.sol";
+
+/**
+ * @dev Alternative to {Nonces}, that support key-ed nonces.
+ *
+ * Follows the https://eips.ethereum.org/EIPS/eip-4337#semi-abstracted-nonce-support[ERC-4337's semi-abstracted nonce system].
+ */
+abstract contract NoncesSemiAbstracted is Nonces {
```

Comment:
This contract completely changes the logic and storage so I don't think it should inherit and override. It should work as a drop-in replacement if someone wants to swap out `Nonces` for this one.

```suggestion
abstract contract NoncesSemiAbstracted {
```